顧客各種画面/受注一覧/メニュー/ヘッダー　レイアウト修正

b25a0031 opened the pull request:
* レイアウト修正
* importの順番を修正

b25a0031 committed:
各種画面レイアウト修正

* Table.vue
* ClientsCreate.vue インポート/レスポンシブ/メッセージ
* ClientsDelete.vue　インポート/レスポンシブ/メッセージ
* ClientsEdit.vue　インポート/レスポンシブ/メッセージ
* ClientsList.vue　インポート/レスポンシブ/メッセージ

Sato0823 committed:
軽微な修正

* importの書き方の統一
* 不要なmsg類の削除
* メッセージの書き方の統一(。を付けるなど)
* 受注一覧テーブルに商品コードを表示するように修正

Sato0823 committed:
データベース制約追加とそれに伴う修正

* index.js138行目付近。受注情報画面に登録されいる顧客情報が削除された際に422レスポンスを返す処理を追加
* init.sqlに外部キー制約を追加する処理を追加
* ClientsDelete.vue166行目付近。index.jsで返されたレスポンスを判別し、エラーメッセージを表示する処理を追加
* 各顧客情報関連のvueファイル。エラーメッセージの表示位置のミス、不要なタグの記載があったので、修正

b25a0031 committed:
clients.js レビュー指摘反映

* clients.js 134行目　コメントの修正

b25a0031 committed:
ClientsDelete.vue 修正

* ClientsDelete.vue 168行目　削除対象の顧客が受注情報に登録されている場合、OKクリック後、顧客一覧画面に遷移するよう修正

NishidaR commented on front/src/components/Header.vue:
```diff
@@ -40,9 +40,9 @@ export default {
+        this.$router.push({ name: "logIn", params: { flashMsg: "ログアウトしました。" } });
       } catch (e) {
-        this.$router.push({ name: "logIn", params: { flashErrMsg: "ログアウト中にエラーが発生しました" } });
+        this.$router.push({ name: "logIn", params: { flashErrMsg: "ログアウト中にエラーが発生しました。" } });
```
詳細設計側で指摘したのにコード側で指摘漏れていました・・。
詳細設計書レビュー指摘票のNo.67の通り、`UserUtil.deleteCurrentUserInfo();`でエラーになることはないと思います。
なので、try/catchは不要ですし、catch側で設定しているメッセージがログイン画面に表示されることはない認識です。
※詳細設計側でもこのメッセージは設計していないはずです。

NishidaR commented on backend/index.js:
```diff
@@ -135,8 +135,13 @@ app.delete("/api/clients/:clientNo", async function (req, res) {
     //異常レスポンス
-    console.log("failed to delete client", e);
-    res.status(500).send("server error occur");
+    if (e.parent.errno == 1451) {
```
調査依頼です。
`1451`はsequelizeが出力するエラーコードだと思っていますが、sequelizeのモジュールからimportすることって可能そうでしょうか？
可能であればimportしてそれを使うように修正いただきたく。（無理なら無理でいいので調査は最大でも15分で見切りをつけてください。）

Sato0823 replied:
調査の結果、エラーコード1451か否かの判定は難しそうだったので、
発生したエラーeがsequelize.ForeignKeyConstraintErrorのインスタンスかどうかを判定する方法に切り替えました。
以下のコードのようになりますが、このような修正で進めてしまって問題ないでしょうか？
(簡単に動かしてみたところ、正常に動作します)

```
const sequelize = require("sequelize");

catch (e) {
    //異常レスポンス
    if (e instanceof sequelize.ForeignKeyConstraintError) {
      // 受注情報が削除しようとしてる顧客番号を参照している場合
      res.status(422).send("unable to delete");
    } else {
      console.log("failed to delete orders.", e);
      res.status(500).send("server error occur");
    }
  }
```

NishidaR replied:
依頼内容がよくなかったですね・・。
まず検証方法は元のエラーコードでの比較でOKです。

依頼したかったのは、もし1451というコードが定数かなにかでsequelize側で定義されていれば、それをimportできないかを調べてほしかったです。
第三者は1451という値が何を意味するのかがわからないので、意味が解る状態にしたいのが意図です。
自力で変数を作ってもいいですが、もしsequelize側で定義を持ってるなら、それを流用したほうが確実だし楽だよね。って感じです。

Sato0823 replied:
ご回答ありがとうございます。

初めは依頼通りに、seaqlizeから1451というエラーコードを引っ張ってこれないかを検討していたのですが、
それが厳しそうだったので、「エラーコード1451か否かの判定は難しそう」という旨をお伝えしました。
誤解を招く記述でした、、日本語が不十分で申し訳ございません、、

仰るように、1451がいきなり出てくる不自然さは自分も感じていたため、
seaqlizeのエラークラスで判定する上記のようなコードを記述させていただきました。

もし、エラーコードでの比較を行うのであれば、
seaqlizeで定数を取得できない(自分が調べた限りでは難しそうでした、、)ので、
以下のような修正になりそうですが如何でしょうか。
```

/**
 * 顧客情報削除API
 */
app.delete("/api/clients/:clientNo", async function (req, res) {
  // 外部キー制約に反する場合のエラーコード
  const ForeignKeyConstraintErrorCode = 1451;
  try {
    await ClientsLogic.delete(db, req.params.clientNo);
    //正常レスポンス
    res.send();
  } catch (e) {
    //異常レスポンス
    if (e.parent.errno == ForeignKeyConstraintErrorCode) {
      // 受注情報が削除しようとしてる顧客番号を参照している場合
      res.status(422).send("unable to delete");
    } else {
      console.log("failed to delete orders.", e);
      res.status(500).send("server error occur");
    }
  }
});

```

NishidaR replied:
1451はsequelizeが決めているエラーコードではなくmysql標準のエラーコードのようで、sequelizeからは拾えなさそうでした・・。

大枠はご提案いただいている実装でOKですが、このエラーコードは顧客情報削除固有ではなく汎用的であるべきなので、ローカル変数として定義するのは微妙です。
最悪フィールド変数でもいいですが、それも若干微妙なのでユーティリティに持たせる方針でどうでしょうか？

◆イメージ
/db/utility.js（データベースのユーティリティ）
```
module.exports.ErrorCode = {
  foreignKeyConstraint: 1451
}
```

/index.js
`if (e.parent.errno == DbUtil.ErrorCode.foreignKeyConstraint) {`

Sato0823 replied:
お忙しい中、ご回答いただきありがとうございます。
/db/utility.jsに持たせる方法で実装できそうなので、
その方向で進めさせていただきます。

NishidaR commented on front/src/views/clients/ClientsCreate.vue:
```diff
@@ -264,15 +236,23 @@ export default {
           alert("顧客情報登録に失敗しました。");
+          console.log(e);
         }
-        console.log(e);
```
質問です。
スタックトレースのログ出力をif/elseの中にそれぞれ書くようにした理由を教えてください。

b25a0031 replied:
申し訳ございません。レイアウト修正の際に誤ってレビュー指摘反映前のものをコピペしてしまいました。

NishidaR replied:
誤りということであれば修正お願いします。

NishidaR commented on front/src/views/clients/ClientsCreate.vue:
```diff
@@ -264,15 +236,23 @@ export default {
 };
 </script>
+<style scoped>
+@media screen and (min-width: 992px) {
```
質問です。
このcssはどういった意図で実装して、どのような動作をしますか？

b25a0031 replied:
画面サイズを変更したときにエラーメッセージがテキストボックスに合わせて改行するのを防ぐためのものと聞いております。（横田君実装の為、修正が難しいので不要であれば削除いたします。）

NishidaR replied:
検討します。こちらの件はひとまずスキップで大丈夫です。

NishidaR replied:
あってもなくても違いがよくわからなかったですが、ひとまずこのままでOKです。
以降の画面作成では、必要であれば追加としてください。

NishidaR commented on front/src/views/clients/ClientsDelete.vue:
```diff
@@ -112,13 +100,16 @@ export default {
       }
 
+      // メッセージ設定
+      this.errMsg = "";
```
この初期化はいらないと思いました。
mountedの呼び出しタイミングで、`errMsg`初期化時の値（data()内で定義している値）から変更されることはないと思ったためです。

NishidaR commented on front/src/views/clients/ClientsDelete.vue:
```diff
@@ -174,7 +163,12 @@ export default {
-        window.alert("顧客情報削除処理に失敗しました。");
+        if (e.response.status === 422) {
+          window.alert("受注情報に登録されている顧客のため、削除できません。");
+          this.$router.push({ name: "clientsList" });
```
質問です。
外部キー制約エラーとなった場合に顧客情報一覧画面に遷移するようにした意図を教えてください。

b25a0031 replied:
受注情報を削除しない限り削除できない為、その場に留まり続けても意味がないと判断し、このような仕様といたしました。

NishidaR replied:
このエラーを受けて、どうしても削除したいから受注情報を削除する人もいれば、仕方ないから諦めて別の顧客情報を削除する人もいると思うので、遷移先がユーザーによって変わるかなって思います。
決めの問題ではありますが、今回は上流工程のどこでも定義してないですし、遷移先がベストというわけでもないのでノータッチでいいと思いましたがいかがでしょうか？
※ちなみに、500エラーもその場に留まり続けても意味がない。ってなるケースがほとんどだと思いますが、特に画面遷移とかはしないので統一感もでないような気がします。

b25a0031 replied:
ご指摘ありがとうございます。仰る通り統一感もなく、ユーザーによってどうしたいのかも変わるため、遷移はしないようにいたします。

NishidaR commented on front/src/views/clients/ClientsList.vue:
```diff
@@ -82,17 +85,17 @@ export default {
       await this.getClients();
     } catch (e) {
-      this.$router.push({ name: "logIn", params: { flashMsg: "ログインしてください" } });
+      this.$router.push({ name: "logIn", params: { flashMsg: "ログインしてください。" } });
```
`serUtil.isLogIn()`とか`this.getClients()`の関数呼び出しで例外スローされた場合に呼ばれると思いますが、ログインしていないことが原因であることはないと思います。
おそらくこの処理は設計に登場しないと思っていますが、そこまで修正しだすと時間が勿体ないので設計の修正はなしでテストも不要としてください。
他画面ではここの処理がバラバラなので、ひとまず`this,errMsg = e.message`に統一するようにしてください。（レビュー漏れてましたね…すみません…）
※本処理でエラーになるようであれば再ログインさせても仕方ないため。

NishidaR commented on front/src/views/orders/OrdersList.vue:
```diff
@@ -82,37 +77,51 @@ export default {
       try {
         const response = await AjaxUtil.getOrders();
         const tmpResponse = JSON.parse(response.data.Items);
+        // 配列に入っている値を一つずつ取り出し、新しい変数を追加していく処理
```
コメント間違っていると思います。

間違っていないのであれば、
`新しい変数を追加していく`の、「新しい変数」はなんの変数か？「追加していく」はどこに追加？かを返信ください。

b25a0031 replied:
申し訳ございません。コピペミスだと考えられますので修正いたします。

NishidaR commented on sql/init.sql:
```diff
@@ -3,6 +3,8 @@ USE sales;
 #------------------------------------------
 # ユーザー情報TBL 作成
 #------------------------------------------
+DROP TABLE IF EXISTS orders;
```
ユーザーテーブルではないのでここに書くべき処理ではないと思います。

Sato0823 replied:
申し訳ございません。
この記述に関しては、再度init.sqlを実行しようとした際に、
外部キー制約があるため、
子テーブルであるordersテーブルが存在している状態だと、
親テーブルであるproducts、clientsテーブルを削除できないという問題が発生したため追加しました。
記述位置の変更や、コメントの記載を行いたいと思います。

NishidaR replied:
記述位置の変更で大丈夫だと思います。元からその順番で書いてあれば不思議に思う人もいないと思います。
※勝手に誰かが順番変えてもエラーがでれば外部キー制約が原因なのはわかると思いますので。

Sato0823 replied:
ご指摘ありがとうございます。
記述位置の変更のみを行いたいと思います。

Sato0823 committed:
レビュー指摘反映1

* index.js98行目。エラーメッセージの修正。
* index.js138行目付近。エラーコードをdb/utility.jsから取得するように変更
* utility.js19行目～。エラーコードを持つ連想配列を追加し、外部キー制約違反のエラーコードを追加
* Header.vue41行目付近。tyrcatch文を削除。
* ClientsCreate.vue239行目付近。エラーを出力するconsole.log()をifelse文の外側に移動させ、冗長にならないように修正。
* ClientsCreate.vue252行目。コメント追加。
ClientsDelete.vue104行目付近。権限チェックは他のところでは行っていないので削除し、それに伴い81行目のUserConstのインポートの記述を削除。
* ClientsCreate.vue110,111行目。メッセージの初期化する処理を削除。
* OrdersList.vue118行目。コメントを削除。
* init.sql3行目。ordersテーブルを削除する記述の位置を変更。

NishidaR commented on sql/init.sql:
```diff
@@ -1,9 +1,10 @@
 USE sales;
 
+DROP TABLE IF EXISTS orders;
```
受注情報だけ浮くので、以下の感じで全体を修正可能ですか？

１．各テーブル削除（DROP TABLE）
２．各テーブル作成（CREATE TABLE）
３．各テーブル初期データ投入（INSERT）

ついでですみません。
３番の初期データ導入前にDELETEしていますが、１番でDROPしているので不要になります。
DELETE分は一律削除お願いします。

NishidaR commented on backend/db/utility.js:
```diff
@@ -16,3 +16,10 @@ module.exports.connect = function () {
+ */
+module.exports.ErrorCode = {
+  foreignKeyConstraint: 1451
+};
```
フォーマットしたら末尾に空行入った気がするので再フォーマットかけてみてください。
再フォーマットしても変更ないということであれば、空行は不要なので無視してください。

Sato0823 committed:
レビュー指摘反映2

* init.sql:可読性工場の為、処理の順番を変更
* db/utility.js:フォーマッターを実行

NishidaR commented on backend/index.js:
```diff
@@ -135,8 +135,13 @@ app.delete("/api/clients/:clientNo", async function (req, res) {
+      // 受注情報が削除しようとしてる顧客番号を参照している場合
+      res.status(422).send("unable to delete");
+    } else {
+      console.log("failed to delete order.", e);
```
受注情報に依存していますが、受注情報の削除に失敗したわけではないと思います。

NishidaR commented on front/src/views/orders/OrdersList.vue:
```diff
@@ -151,8 +168,6 @@ export default {
       this.$router.push({ name: "ordersDelete", query: { orderNo: this.orderRow.order_no } });
     },
-
-    deliveryNoteOutput() {},
```
納品書出力ボタン押下時にエラーとなるため、関数を戻すか、ボタン押下時のイベントを削除してください。

NishidaR commented on sql/init.sql:
```diff
@@ -1,9 +1,17 @@
+#------------------------------------------
+# ユーザー情報TBL 作成
+#------------------------------------------
+
```
細かいですが、気になったので空行削除お願いします。

Sato0823 committed:
レビュー指摘反映3

* index.js142行目:コンソールで出力するエラーメッセージのミスを修正
* OrdersList.vue172行目～:納品書出力処理の関数を追加
* init.sql14行目:空行を削除
